tst/performance/test_container_iterator.cpp

```diff
@@ -0,0 +1,282 @@
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#include <array>
+#include <cmath>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <catch2/catch.hpp>
+
+#include "athena.hpp"
+#include "basic_types.hpp"
+#include "interface/container.hpp"
+#include "interface/container_iterator.hpp"
+#include "interface/metadata.hpp"
+#include "interface/variable.hpp"
+#include "interface/variable_pack.hpp"
+#include "kokkos_abstraction.hpp"
+#include "parthenon_arrays.hpp"
+
+using parthenon::CellVariable;
+using parthenon::CellVariableVector;
+using parthenon::Container;
+using parthenon::ContainerIterator;
+using parthenon::DevExecSpace;
+using parthenon::loop_pattern_mdrange_tag;
+using parthenon::Metadata;
+using parthenon::MetadataFlag;
+using parthenon::PackIndexMap;
+using parthenon::par_for;
+using parthenon::ParArray4D;
+using parthenon::ParArrayND;
+using parthenon::Real;
+using parthenon::X1DIR;
+using parthenon::X2DIR;
+using parthenon::X3DIR;
+
+// Test wrapper to run a function multiple times
+template <typename InitFunc, typename PerfFunc>
+double performance_test_wrapper(const int n_burn, const int n_perf, InitFunc init_func,
+                                PerfFunc perf_func) {
+  // Initialize the timer and test
+  Kokkos::Timer timer;
+  init_func();
+
+  for (int i_run = 0; i_run < n_burn; i_run++) {
+    // burn
+    perf_func();
+  }
+  Kokkos::fence();
+  timer.reset();
+  for (int i_run = 0; i_run < n_perf; i_run++) {
+    // time
+    perf_func();
+  }
+
+  // Time it
+  Kokkos::fence();
+  double perf_time = timer.seconds();
+
+  // FIXME?
+  // Validate results?
+
+  return perf_time;
+}
+
+TEST_CASE("Container Iterator Performance", "[ContainerIterator][performance]") {
+  const int N = 32; // Dimensions of blocks
+  const int Nvar = 10;
+  const int n_burn = 500; // Num times to burn in before timing
+  const int n_perf = 500; // Num times to run while timing
+
+  // Make a raw ParArray4D for closest to bare metal looping
+  ParArrayND<Real> raw_array("raw_array", Nvar, N, N, N);
+
+  // Make a function for initializing the raw ParArray4D
+  auto init_raw_array = [&]() {
+    par_for(
+        "Initialize ", DevExecSpace(), 0, Nvar - 1, 0, N - 1, 0, N - 1, 0, N - 1,
+        KOKKOS_LAMBDA(const int l, const int k, const int j, const int i) {
+          raw_array(l, k, j, i) =
+              static_cast<Real>((l + 1) * (k + 1) * (j + 1) * (i + 1));
+        });
+  };
+
+  // Test performance iterating over variables (we should aim for this performance)
+  double time_raw_array = performance_test_wrapper(n_burn, n_perf, init_raw_array, [&]() {
```

Comment:
How easy would it be to, essentially, replace these `performance_test_wrapper` calls with Catch2's benchmarking macros: https://github.com/catchorg/Catch2/blob/master/docs/benchmarks.md#top

Reply:
You'd need to make sure to keep the fences, but you could get rid of the timers.

Reply:
> How easy would it be to, essentially, replace these `performance_test_wrapper` calls with Catch2's benchmarking macros

I will look into it next week.

Reply:
> You'd need to make sure to keep the fences, but you could get rid of the timers.

Speaking of fences, I'm wondering if we actually even need more fences and/or add an option to fence between the executions within the `n_perf` loop (basically allowing to include the individual launch overhead in the timing or not).

Reply:
It would be useful as an option! But we definitely want to have a performance test that minimizes the overhead.

Reply:
We can create an [issue](https://github.com/lanl/parthenon/issues/198) for this.